src/asl.ts

```diff
@@ -2837,60 +2792,494 @@ export class ASL {
+        return value ? `${value}.${expr.name.name}` : undefined;
       } else if (isElementAccessExpr(expr)) {
-        return `${toFilterCondition(
-          expr.expr
-        )}[${this.assertElementAccessConstant(
+        const field = this.assertElementAccessConstant(
           ASLGraph.getAslStateOutput(this.eval(expr.element))
-        )}]`;
+        );
+
+        const value = toFilterCondition(expr.expr);
+
+        return value
+          ? `${value}[${typeof field === "number" ? field : `'${field}'`}]`
+          : undefined;
       }
 
+      // try to compile with ASLGraph instead
+      return undefined;
+    };
+
+    const expression = toFilterCondition(stmt.expr);
+    return expression
+      ? {
+          jsonPath: `${valueJsonPath.jsonPath}[?(${expression})]`,
+        }
+      : undefined;
+  }
+
+  /**
+   * filter an array using ASLGraph. Should support any logic that returns a boolean already supported by Functionless Step Functions.
+   *
+   * Note: If possible, try to use {@link filterToJsonPath} as it introduces no new state transitions.
+   */
+  private filterToASLGraph(
+    valueJsonPath: ASLGraph.JsonPath,
+    predicate: FunctionExpr | ArrowFunctionExpr
+  ): ASLGraph.OutputSubState {
+    const predicateResult = this.newHeapVariable();
+    const filterResult = this.newHeapVariable();
+
+    const [itemParameter, indexParameter, arrayParameter] =
+      predicate.parameters;
+
+    const assignParameters = ASLGraph.joinSubStates(
+      predicate,
+      itemParameter
+        ? this.evalDecl(itemParameter, {
+            jsonPath: indexParameter
+              ? `${filterResult}.arr[0].item`
+              : `${filterResult}.arr[0]`,
+          })
+        : undefined,
+      indexParameter
+        ? this.evalDecl(indexParameter, {
+            jsonPath: `${filterResult}.arr[0].index`,
+          })
+        : undefined,
+      arrayParameter
+        ? this.evalDecl(arrayParameter, {
+            jsonPath: valueJsonPath.jsonPath,
+          })
+        : undefined
+    );
+
+    const uniqueReturnName = this.generatedNames.generateOrGet(predicate);
+
+    const predicateBody = this.evalStmt(predicate.body, {
+      End: undefined,
+      ResultPath: predicateResult,
+      Next: uniqueReturnName,
+    });
+
+    return {
+      startState: "init",
+      states: {
+        init: indexParameter
+          ? {
+              Type: "Map",
+              ItemsPath: valueJsonPath.jsonPath,
+              Parameters: {
+                "index.$": "$$.Map.Item.Index",
+                "item.$": "$$.Map.Item.Value",
+              },
+              Iterator: this.aslGraphToStates({
+                Type: "Pass",
+                ResultPath: "$",
+                Next: ASLGraph.DeferNext,
+              }),
+              ResultSelector: {
+                "arr.$": "$",
+                arrStr: "[null",
+              },
+              ResultPath: filterResult,
+              Next: "check",
+            }
```

Comment:
So if you use index, then you can't mutate a variable outside the scope?

Reply:
No. That map just zips up the index and the item. If you use index you get more state transitions.